src/disp_s1/validate.py

```diff
@@ -487,3 +523,201 @@ def compare(golden: Filename, test: Filename, data_dset: str = DSET_DEFAULT) ->
+
+    """
+    logger.info("Checking for congruence with wrapped phase...")
+
+    # Get unwrapped phase and connected component label data from the product.
+    with h5py.File(product_file, mode="r") as f:
+        unw = f["unwrapped_phase"][()]
+        conncomp = f["connected_component_labels"][()]
+
+    # Get a mask of valid pixels (that had nonzero CC label) and NaN-valued pixels.
+    valid_mask = conncomp != 0
+    nan_mask = np.isnan(unw)
+
+    # Get wrapped phase data, in radians.
+    logger.info(f"Interferogram file: {igram_file}")
+    igram = io.load_gdal(igram_file)
+    wrapped = np.angle(igram)
+
+    # Check that the unwrapped phase is congruent with the wrapped phase.
+    _check_phase_congruence(
+        unw=unw,
+        ref=wrapped,
+        mask=(valid_mask & ~nan_mask),
+        atol=atol,
+    )
+
+
+def _get_frame_id(hdf5_file: Filename) -> int:
+    """Get the frame ID of an OPERA DISP-S1 product.
+
+    Parameters
+    ----------
+    hdf5_file : Filename
+        The product file path.
+
+    Returns
+    -------
+    int
+        The frame ID number of the product.
+
+    """
+    with h5py.File(hdf5_file, mode="r") as f:
+        return f["identification/frame_id"][()]
+
+
+def _check_frame_bounds(
+    filename: Filename,
+    frame_id: int,
+    json_file: Filename | None = None,
+    *,
+    atol: float = 1e-6,
+) -> None:
+    """Validate a dataset's spatial reference and bounding box.
+
+    Compare the EPSG code and spatial extents of the product with the expected bounds
+    from the frame-to-burst JSON file.
+
+    Parameters
+    ----------
+    filename : Filename
+        The file path of the raster dataset to validate.
+    frame_id : int
+        The frame ID number of the dataset.
+    json_file : Filename or None, optional
+        The file path of the frame-to-burst JSON file. If None, uses the vendored
+        frame-to-burst file. Defaults to None.
+    atol : float, optional
+        The absolute tolerance used for comparing bounding box coordinates. Must be
+        nonnegative. Defaults to 1e-6.
+
+    Raises
+    ------
+    ValidationError
+        If the EPSG code of the raster dataset did not match the EPSG code of the
+        corresponding frame in the JSON file.
+    ValidationError
+        If the bounding box of the raster dataset did not match the bounding box of the
+        corresponding frame in the JSON file.
+
+    """
+    logger.info("Checking frame bounds again JSON file...")
+
+    if atol < 0.0:
+        errmsg = f"atol must be >= 0, got {atol}"
+        raise ValueError(errmsg)
+
+    if json_file is None:
+        json_file = FRAME_TO_BURST_JSON_FILE
+
+    logger.info(f"Dataset name: {filename}")
+    logger.info(f"Frame ID: {frame_id}")
+    logger.info(f"Frame-to-burst JSON file: {json_file}")
+
+    # Extract the EPSG code and bounding box for this frame from the JSON file.
+    json_epsg, json_bbox = get_frame_bbox(frame_id=frame_id, json_file=json_file)
```

Comment:
This has just reminded me that there's the redundant `get_frame_bbox`/other functions in disp-s1 that are now all in `opera-utils`... we should probably switch those over at some point

Reply:
(solved in https://github.com/opera-adt/disp-s1/pull/89)